src/traces/contourcarpet/plot.js

```diff
@@ -0,0 +1,305 @@
+        .classed('contourlevel', true);
+    linegroup.exit().remove();
+
+    var opencontourlines = linegroup.selectAll('path.openline')
+        .data(function(d) { return d.pedgepaths; });
+    opencontourlines.enter().append('path')
+        .classed('openline', true);
+    opencontourlines.exit().remove();
+    opencontourlines
+        .attr('d', function(d) {
+            return Drawing.smoothopen(d, smoothing);
+        })
+        .style('vector-effect', 'non-scaling-stroke');
+
+    var closedcontourlines = linegroup.selectAll('path.closedline')
+        .data(function(d) { return d.ppaths; });
+    closedcontourlines.enter().append('path')
+        .classed('closedline', true);
+    closedcontourlines.exit().remove();
+    closedcontourlines
+        .attr('d', function(d) {
+            return Drawing.smoothclosed(d, smoothing);
+        })
+        .style('vector-effect', 'non-scaling-stroke')
+        .style('stroke-miterlimit', 1);
+}
+
+function makeBackground(plotgroup, clipsegments, xaxis, yaxis, isConstraint, coloring) {
+    var seg, xp, yp, i;
+    var bggroup = makeg(plotgroup, 'g', 'contourbg');
+
+    var bgfill = bggroup.selectAll('path')
+        .data((coloring === 'fill' && !isConstraint) ? [0] : []);
+    bgfill.enter().append('path');
+    bgfill.exit().remove();
+
+    var segs = [];
+    for(i = 0; i < clipsegments.length; i++) {
+        seg = clipsegments[i];
+        xp = map1dArray([], seg.x, xaxis.c2p);
+        yp = map1dArray([], seg.y, yaxis.c2p);
+        segs.push(makepath(xp, yp, seg.bicubic));
+    }
+
+    bgfill
+        .attr('d', 'M' + segs.join('L') + 'Z')
+        .style('stroke', 'none');
+}
+
+function makeFills(trace, plotgroup, xa, ya, pathinfo, perimeter, ab2p, carpet, carpetcd, coloring, boundaryPath) {
+    var fillgroup = plotgroup.selectAll('g.contourfill')
+        .data([0]);
+    fillgroup.enter().append('g')
+        .classed('contourfill', true);
+
+    var fillitems = fillgroup.selectAll('path')
+        .data(coloring === 'fill' ? pathinfo : []);
+    fillitems.enter().append('path');
+    fillitems.exit().remove();
+    fillitems.each(function(pi) {
+        // join all paths for this level together into a single path
+        // first follow clockwise around the perimeter to close any open paths
+        // if the whole perimeter is above this level, start with a path
+        // enclosing the whole thing. With all that, the parity should mean
+        // that we always fill everything above the contour, nothing below
+        var fullpath = joinAllPaths(trace, pi, perimeter, ab2p, carpet, carpetcd, xa, ya);
+
+        if(pi.prefixBoundary) {
+            fullpath = boundaryPath + fullpath;
+        }
+
+        if(!fullpath) {
+            d3.select(this).remove();
+        } else {
+            d3.select(this)
+                .attr('d', fullpath)
+                .style('stroke', 'none');
+        }
+    });
+}
+
+/* function clipGaps(plotGroup, plotinfo, cd0, perimeter) {
```

Comment:
Yup. TODO. I'll remove it. It's in the git history or just copied from contour anyway.